src/modules/paymentProcessor/PP_Queue_v1.sol

```diff
@@ -0,0 +1,941 @@
+
+        _claimPreviouslyUnclaimable(client, token, receiver);
+    }
+
+    /// @inheritdoc IPaymentProcessor_v1
+    function validPaymentOrder(
+        IERC20PaymentClientBase_v1.PaymentOrder memory order_
+    ) external view returns (bool isValid_) {
+        // Extract queue ID from order data.
+        (uint queueId_,,,) = _getPaymentQueueDetails(order_.flags, order_.data);
+
+        // Validate payment receiver, amount and queue ID.
+        isValid_ = _validPaymentReceiver(order_.recipient)
+            && _validTotal(order_.amount) && _validQueueId(queueId_);
+    }
+
+    /// @inheritdoc IPP_Queue_v1
+    function addPaymentOrderToQueue(
+        address client_,
+        address token_,
+        address receiver_,
+        uint amount_,
+        bytes calldata data_
+    ) external onlyModuleRole(QUEUE_OPERATOR_ROLE) returns (uint orderId_) {
+        bytes32[] memory decodedData = new bytes32[](data_.length / 32);
+        assembly {
+            calldatacopy(
+                add(decodedData, 32),
+                add(data_.offset, 32),
+                mul(mload(decodedData), 32)
+            )
+        }
+
+        // Check token balance and allowance.
+        if (!_validTokenBalance(token_, client_, amount_)) {
+            revert Module__PP_Queue_QueueOperationFailed(client_);
+        }
+
+        // Create payment order.
+        IERC20PaymentClientBase_v1.PaymentOrder memory order =
+        IERC20PaymentClientBase_v1.PaymentOrder({
+            recipient: receiver_,
+            paymentToken: token_,
+            amount: amount_,
+            originChainId: block.chainid,
+            targetChainId: block.chainid,
+            flags: 0,
+            data: decodedData
+        });
+
+        // Add to queue.
+        orderId_ = _addPaymentOrderToQueue(order, client_);
+    }
+
+    /// @inheritdoc IPP_Queue_v1
+    function cancelPaymentOrderThroughQueueId(uint orderId_)
+        external
+        onlyModuleRole(QUEUE_OPERATOR_ROLE)
+        returns (bool success_)
+    {
+        // Validate queue ID.
+        if (!_orderExists(orderId_)) {
+            revert Module__PP_Queue_InvalidOrderId(
+                _orders[orderId_].client, orderId_
+            );
+        }
+
+        QueuedOrder storage order = _orders[orderId_];
+
+        // Check if order can be cancelled
+        if (order.state != RedemptionState.PROCESSING) {
+            revert Module__PP_Queue_InvalidState();
+        }
+
+        // Update order state and emit event
+        _updateOrderState(orderId_, RedemptionState.CANCELLED);
+
+        // Add amount to unclaimable
+        _setUnclaimableAmount(
+            _msgSender(),
+            order.order.paymentToken,
+            order.order.recipient,
+            order.order.amount
+        );
+
+        // Remove from queue if present
+        _removeFromQueue(orderId_);
+
+        success_ = true;
+    }
+
+    /// @inheritdoc IPP_Queue_v1
+    function processNextOrder(address client_)
```

Comment:
Why do we need this external function ser

Reply:
mmmmm... I guess we're good with the internal one then